crates/scheduler/src/batch_scheduler.rs

```diff
@@ -0,0 +1,209 @@
+            }
+            progress::Progress::Status(progress::Status {
+                batch_size,
+                round_time,
+            }) => {
+                let mut tracker = tracker.lock().await;
+                tracker.update(&peer_id, round_time, batch_size)?;
+                match tracker.worker_tracker.worker_state(&peer_id) {
+                    State::Training => {
+                        let time_cap = 10000u64;
+                        let update_cap = 3u64;
+                        let (time, cnt, projection) = simulator(
+                            tracker.worker_tracker.last_updates(),
+                            tracker.worker_tracker.batch_sizes(),
+                            tracker.worker_tracker.estimates(),
+                            tracker.count(),
+                            time_cap,
+                            update_cap,
+                        );
+                        let peer_position = tracker.worker_tracker.worker_position(&peer_id)?;
+                        tracing::debug!(time = %time, count = %cnt, "Finish simulation");
+                        tracing::info!(
+                            "Simulation with projection {:?} and {:?} of {:?}",
+                            projection,
+                            cnt,
+                            tracker.count()
+                        );
+                        if cnt == 0 && projection[peer_position] < update_cap {
+                            tracker
+                                .worker_tracker
+                                .update_worker_state(&peer_id, State::UpdateScheduled);
+                            Ok(progress::Response::ScheduleUpdate {
+                                counter: projection[peer_position],
+                            })
+                        } else {
+                            Ok(progress::Response::Continue {})
+                        }
+                    }
+                    State::UpdateScheduled => Ok(progress::Response::Continue {}),
+                    State::Done => Ok(progress::Response::Done {}),
+                    _ => Err(BatchSchedulerError::Scheduling("Wrong state".into())),
+                }
+            }
+            progress::Progress::Update => {
+                tracker
+                    .lock()
+                    .await
+                    .worker_tracker
+                    .update_worker_state(&peer_id, State::Updating);
+                Ok(progress::Response::Ok {})
+            }
+            progress::Progress::Updated => {
+                let mut tracker = tracker.lock().await;
+                tracker.next_round();
+                if tracker.training_finished() {
+                    tracker.worker_tracker.done();
+                    Ok(progress::Response::Done {})
+                } else {
+                    Ok(progress::Response::Ok {})
+                }
+            }
+        }
+    }
+
+    pub async fn run(
```

Comment:
Something I used in #102 and which would make sense here as well: Have `BatchScheduler` implement `Stream` instead of having this function. The `handle` function becomes part of the struct and is therefore tied to the lifetime of a `BatchScheduler` instance. As long as the instance is in scope, it'll handle requests.

Furthermore, with the current `run` function, we wouldn't need a `BatchScheduler` instance... all of the state held in this instance is only used in this function. I.e., we could also simply have a function

```
run(network: Network, simulator: SimulatorFn, tracker: Arc<Mutex<ProgressTracker<T>>>, job_id: Uuid)
```